### PR TITLE
feat: add ray plugin for job

### LINKED ISSUE #4182: Volcano job natively support Ray framework
### What is the problem you're trying to solve

Ray is a popular AI framework that has been widely used, and ray operator has already supported Volcano as a batch scheduler, see：https://github.com/ray-project/kuberay
On the other hand, many users still use volcano job to run their job instead of new ray API，so it's reasonable to support ray natively on volcano job.

### Describe the solution you'd like

Currently volcano supports distributed AI and HPC framework like pytorch, tensorflow,  mpi.
Just like what vcjobs currently do, and a new plugin named `ray` in pkg/controllers/job/plugins/distributed-framework, so users can submit a vcjob and actually run a ray job.

### PR DESCRIPTION
**Pull request by Wonki4**
#### What type of PR is this?
This PR includes the ray plugin support for volcano job.

#### What this PR does / why we need it:
This PR supports users to build a ray cluster is composed of head and worker easily.

Example

```yaml
--- # without ray plugin
apiVersion: batch.volcano.sh/v1alpha1
kind: Job
metadata:
  name: ray-cluster-job
spec:
  minAvailable: 3
  schedulerName: volcano
  plugins:
    svc: []
  queue: default
  policies:
    - event: PodEvicted
      action: RestartJob
  tasks:
    - replicas: 1
      name: head
      template:
        spec:
          containers:
            - name: head
              command:
                - sh
                - -c
                - ray start --head --block --port=6379 --dashboard-host=0.0.0.0;
              image: rayproject/ray:latest-py311-cpu
              ports:
                - containerPort: 8265
                  name: dashboard
                - containerPort: 6379
                  name: gcs
                - containerPort: 10001
                  name: client
              resources: {}
          restartPolicy: OnFailure
    - replicas: 2
      name: worker
      template:
        spec:
          containers:
            - name: worker
              command:
                - sh
                - -c
                - |
                  ray start --block --address=ray-cluster-job-head-0.ray-cluster-job:6379
              image: rayproject/ray:latest-py311-cpu
              resources: {}
          restartPolicy: OnFailure 
```

```yaml
--- # with ray plugin
apiVersion: batch.volcano.sh/v1alpha1
kind: Job
metadata:
  name: ray-cluster-job
spec:
  minAvailable: 3
  schedulerName: volcano
  plugins:
    ray: []
    svc: []
  policies:
    - event: PodEvicted
      action: RestartJob
  queue: default
  tasks:
    - replicas: 1
      name: head

      template:
        spec:
          containers:
            - name: head
              image: rayproject/ray:latest-py311-cpu
              resources: {}
          restartPolicy: OnFailure
    - replicas: 2
      name: worker
      template:
        spec:
          containers:
            - name: worker
              image: rayproject/ray:latest-py311-cpu
              resources: {}
          restartPolicy: Never
```

If a user use ray plugin, the cluster configuration is easier than before
- Head Node Command Setup
- Worker Node Command Setup
- Port Setup (gcs, dashboard and client server)
- How to configure a ray cluster
- Service connected to a head task pod.

Here is the ray plugin arguments
- head: The name of a head task
- headContainer: The container name in a head task
- worker: The name of a worker task
- workerContainer: The container name in a worker task
- port: GCS Port
- dashboardPort: Ray cluster dashboard port
- clientServerPort: client server port

The architecture of ray cluster in Volcano Job

ray_plugin_diagram drawio

#### Which issue(s) this PR fixes:

Fixes #4182

#### Special notes for your reviewer:
**I know that there is another PR about this issue(#4193 ).**
but there's no progress for 2 months so I decided to start working on it.

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/controllers/job/plugins/distributed-framework/ray/ray.go [hunk 1]
```diff
@@ -0,0 +1,288 @@
+/*
+Copyright 2025 The Volcano Authors.
```
[the remark this answers is not in the document]
Comment (Wonki4) [reply pkg/controllers/job/plugins/distributed-framework/ray/ray.go:2]: As it is 2025, this issue will be resolved.

File: test/e2e/jobseq/ray_plugin.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+			Name: "ray-cluster-job",
+			Min:  1,
+			Policies: []vcbatch.LifecyclePolicy{
+				{
+					Event:  vcbus.PodEvictedEvent,
+					Action: vcbus.RestartJobAction,
+				},
+			},
+			Plugins: map[string][]string{
+				"ray": {"--head=head", "--worker=worker", "--port=2345", "--headContainer=rayproject", "--workerContainer=rayproject"},
```
[the remark this answers is not in the document]
Comment (Wonki4) [reply test/e2e/jobseq/ray_plugin.go]: The container name is based on the container image repository name. so, I set up the container names as a "rayproject". Should I change the way I test?

**Comment by Monokaix** [comment]
> /cc @Monokaix I think we can take on this feature in v1.13 /ok-to-test

+1

**Comment by Monokaix** [comment]
Welcome! Thanks for your contribution.
Please check CI
image

**Comment by Wonki4** [comment]
Okay! I'll fix it!

**Comment by Wonki4** [comment]
@Monokaix 
The problem of ray e2e test was related to the ray image size (about 850mb). I changed the test image and resolved the issue.
But, I just wonder how large the disk size of test linux instance is?

**Comment by kingeasternsun** [comment]
Thanks for your contribution.

File: pkg/controllers/job/plugins/distributed-framework/ray/ray.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+	jobhelpers "volcano.sh/volcano/pkg/controllers/job/helpers"
+	pluginsinterface "volcano.sh/volcano/pkg/controllers/job/plugins/interface"
+)
+
+const (
+	// RayPluginName is the name of the plugin
+	RayPluginName = "ray"
+	// DefaultHead is the default task name of head node.
+	DefaultHead = "head"
+	// DefaultHeadContainer
```
Comment (Monokaix) [inline pkg/controllers/job/plugins/distributed-framework/ray/ray.go]: Seems incomplete.

File: pkg/controllers/job/plugins/distributed-framework/ray/ray.go [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+	DefaultHeadContainer = "head"
+	// DefaultWorker is the default task name of worker node.
+	DefaultWorker = "worker"
+	// DefaultWorkerContainer
```
Comment (Monokaix) [inline pkg/controllers/job/plugins/distributed-framework/ray/ray.go]: same above.

File: pkg/controllers/job/plugins/distributed-framework/ray/ray.go [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+	GcsPortName = "gcs"
+	// DashboardPortName is the port name for a ray dashboard.
+	DashboardPortName = "dashboard"
+	// ClientServerPortName is the port name for a ray client api
+	ClientServerPortName = "client-server"
+)
+
+type rayPlugin struct {
+	rayArguments        []string
+	Clientset           pluginsinterface.PluginClientset
```
Comment (Monokaix) [inline pkg/controllers/job/plugins/distributed-framework/ray/ray.go]:
```suggestion
	clientSet           pluginsinterface.PluginClientset
```

**Comment by Monokaix** [comment]
Please also add the yaml example into `example/integrations`: )

**Comment by Monokaix** [comment]
Can you also check about the functionalities and API fields we missed and can be added into this plugin? Seems it's a little simple, or if it can meet the production requirements under current implementation?

**Comment by Wonki4** [comment]
> Can you also check about the functionalities and API fields we missed and can be added into this plugin? Seems it's a little simple, or if it can meet the production requirements under current implementation?

Okay, I'll check other scenarios using ray. This plugin focus on supporting to create a ray cluster.

**Comment by Monokaix** [comment]
> > Can you also check about the functionalities and API fields we missed and can be added into this plugin? Seems it's a little simple, or if it can meet the production requirements under current implementation?
> 
> Okay, I'll check other scenarios using ray. This plugin focus on supporting to create a ray cluster.

Great, but you can resolve other comments first, we can continuously iterate on this plugin: )

**Comment by Wonki4** [comment]
@Monokaix 
The E2E test was failed due to a disk size.
The ray image is over 400mb at least (the size of ray package is about 300mb) and other images about tensorflow is also big. If the workers of ray and tensorflow are scheduled at the same k8s node, the pod will be pending state because of there are not enough disk size.

Could we increase a disk size of node? or Could you advice me to find a other way to solve it.

> Sep 19 16:33:27 integration-worker3 kubelet[272]: E0919 16:33:27.945147     272 pod_workers.go:1301] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"tensorflow\" with ErrImagePull: \"failed to pull and unpack image \\\"docker.io/volcanosh/dist-mnist-tf-example:0.0.1\\\": failed to extract layer sha256:5061983e267fa42a4403de0a25f41e4ca5ba0c75db3b2a4ceae769f94035816b: write /var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/154/fs/opt/conda/lib/python3.5/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so: no space left on device\"" pod="x2n8bo60/tensorflow-dist-mnist-ps-0" podUID="89d374cc-6b09-4498-a235-163431fcdb72"

**Comment by Monokaix** [comment]
> Sep 19 16:33:27 integration-worker3 kubelet[272]: E0919 16:33:27.945147 272 pod_workers.go:1301] "Error syncing pod, skipping" err="failed to "StartContainer" for "tensorflow" with ErrImagePull: "failed to pull and unpack image \"docker.io/volcanosh/dist-mnist-tf-example:0.0.1\": failed to extract layer sha256:5061983e267fa42a4403de0a25f41e4ca5ba0c75db3b2a4ceae769f94035816b: write /var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/154/fs/opt/conda/lib/python3.5/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so: no space left on device"" pod="x2n8bo60/tensorflow-dist-mnist-ps-0" podUID="89d374cc-6b09-4498-a235-163431fcdb72"

You can use `sudo docker system prune -a -f` or `sudo crictl rmi --prune` to release space before run a e2e.

File: example/integrations/ray/ray-example.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+          restartPolicy: OnFailure
+    - replicas: 2
+      name: worker
+      template:
+        spec:
+          containers:
+            - name: worker
+              image: bitnami/ray:2.49.0
+              resources: {}
+          restartPolicy: Never
```
Comment (Monokaix) [inline example/integrations/ray/ray-example.yaml]: Seems change to `OnFailure` is better, because the worker may cannot connect the header if the header starts after worker.

File: docs/user-guide/how_to_use_ray_plugin.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+| --- | ---------------- | ------ | ------------- | -------- | --------------------------------------- | ------------------------ |
+| 1   | head             | string | head          | No       | Name of Head Task in Volcano Job        | --head=head              |
+| 2   | worker           | string | worker        | No       | Name of Worker Task in Volcano Job      | --worker=worker          |
+| 3   | headContainer    | string | head          | No       | Name of Main Container in a head task   | --headContainer=head     |
+| 4   | workerContainer  | string | worker        | No       | Name of Main Container in a worker task | --workerContainer=worker |
+| 5   | port             | string | 6379          | No       | The port to open for the GCS            | --port=6379              |
+| 6   | dashboardPort    | string | 8265          | No       | The port to open for the Ray dashboard  | --dashboardPort=8265     |
+| 7   | clientServerPort | string | 10001         | No       | The port to open for the client server  | --clientServerPort=10001 |
+
+## Examples
```
Comment (Monokaix) [inline docs/user-guide/how_to_use_ray_plugin.md:33]: Maybe we need also add ray cluster operation doc to users, like https://docs.ray.io/en/master/cluster/kubernetes/getting-started/raycluster-quick-start.html?

Comment (Wonki4) [reply docs/user-guide/how_to_use_ray_plugin.md:33]: Thank you for your check!
Maybe, it will be more comfortable and useful.

Comment (Wonki4) [reply docs/user-guide/how_to_use_ray_plugin.md:33]: Because we suppose that a user doesn't have a ray operator,
I think that below 5 docs are fit for our users.
1. Step 4 ~ Step 5  of [RayCluster Quick Guide](https://docs.ray.io/en/master/cluster/kubernetes/getting-started/raycluster-quick-start.html?) can be useful (Step 1 ~ Step 3 is related to a Ray Operator)
2. What we create and provide is [Ray Cluster Key Concepts](https://docs.ray.io/en/latest/cluster/key-concepts.html)
3. What we based on is [Launching an On-Premise Cluster](https://docs.ray.io/en/latest/cluster/vms/user-guides/launching-clusters/on-premises.html#on-prem)
4. How to use a ray cluster with [Ray Jobs API](https://docs.ray.io/en/latest/cluster/running-applications/job-submission/index.html)
5. With a ray cluster, we can serve a model [Model Deploy on ray cluster](https://docs.ray.io/en/latest/serve/advanced-guides/deploy-vm.html)

Could you comment this? (I think the 5 links are helpful for users to use a ray cluster.)
(If you have, could you tell me a **form** about the ray cluster operation doc?)

File: pkg/controllers/job/plugins/distributed-framework/ray/ray.go (hunk 3 above)
Comment (Wonki4) [reply pkg/controllers/job/plugins/distributed-framework/ray/ray.go]: I filled the annotation.

File: pkg/controllers/job/plugins/distributed-framework/ray/ray.go (hunk 4 above)
Comment (Wonki4) [reply pkg/controllers/job/plugins/distributed-framework/ray/ray.go]: I filled the annotation.

File: pkg/controllers/job/plugins/distributed-framework/ray/ray.go (hunk 5 above)
Comment (Wonki4) [reply pkg/controllers/job/plugins/distributed-framework/ray/ray.go]: I changed the keyword as a lowercase.

File: docs/user-guide/how_to_use_ray_plugin.md (hunk 7 above)
Comment (Monokaix) [reply docs/user-guide/how_to_use_ray_plugin.md:33]: I think it's ok, we have provided a ray cluster and user can run their job, and we just give a simple guide and link the guide doc is ok.

File: example/integrations/ray/ray-example.yaml (hunk 6 above)
Comment (Monokaix) [reply example/integrations/ray/ray-example.yaml]: I have tested locally and the worker failed to start, so we should change to OnFailure here.

**Comment by Monokaix** [comment]
Thanks for your contribution! And we are planning to release new version this week, so it's better to solve all comments to update and merge it today: )

**Comment by Wonki4** [comment]
> Thanks for your contribution! And we are planning to release new version this week, so it's better to solve all comments to update and merge it today: )

Okay! I'll do it!

File: docs/user-guide/how_to_use_ray_plugin.md (hunk 7 above)
Comment (Wonki4) [reply docs/user-guide/how_to_use_ray_plugin.md:33]: The simple guide was integrated into the documentation.

File: example/integrations/ray/ray-example.yaml (hunk 6 above)
Comment (Wonki4) [reply example/integrations/ray/ray-example.yaml]: I changed the restartPolicy. (Never -> OnFailure)

**Comment by Monokaix** [comment]
/approve
Great job!

File: docs/user-guide/how_to_use_ray_plugin.md (hunk 7 above)
Comment (Monokaix) [reply docs/user-guide/how_to_use_ray_plugin.md:33]: 👍

**Comment by JesseStutler** [comment]
/lgtm
Thanks

**Comment by YeonghyeonKO** [comment]
LGTM @Wonki4 Thanks for contributing your code.
